include/deal.II/matrix_free/mapping_info.templates.h

```diff
@@ -2366,114 +2517,44 @@ namespace internal
-                                         cell * dim * n_mapping_points;
-                for (unsigned int d = 0; d < dim; ++d)
-                  {
-                    const double translate_d =
-                      ptr_origin[d * n_mapping_points] -
-                      ptr_mine[d * n_mapping_points];
-                    for (unsigned int q = 1; q < n_mapping_points; ++q)
-                      max_distance =
-                        std::max(std::abs(ptr_origin[d * n_mapping_points + q] -
-                                          ptr_mine[d * n_mapping_points + q] -
-                                          translate_d),
-                                 max_distance);
-                  }
-
-                // this is not a duplicate, must add it again
-                if (max_distance > 1e-10 * jacobian_size)
-                  add_this_cell = true;
-              }
-
-            if (add_this_cell == true)
-              {
-                // check whether cell is Cartesian/affine/general
-                GeometryType type = cartesian;
-                for (unsigned int d = 0; d < dim; ++d)
-                  for (unsigned int e = 0; e < dim; ++e)
-                    if (d != e)
-                      if (std::abs(inserted.first->first[0][d][e]) >
-                          1e-12 * jacobian_size)
-                        type = affine;
-
-                for (unsigned int q = 1; q < n_mapping_points; ++q)
-                  for (unsigned int d = 0; d < dim; ++d)
-                    for (unsigned int e = 0; e < dim; ++e)
-                      if (std::abs(fe_values.jacobian(q)[d][e] -
-                                   fe_values.jacobian(0)[d][e]) >
-                          1e-12 * jacobian_size)
-                        {
-                          type = general;
-                          goto endloop;
-                        }
-              endloop:
-                cell_data_index[cell]       = n_data_buckets;
-                preliminary_cell_type[cell] = type;
-                ++n_data_buckets;
-              }
-            else
-              {
-                cell_data_index[cell] = cell_data_index[inserted.first->second];
-                preliminary_cell_type[cell] =
-                  preliminary_cell_type[inserted.first->second];
-              }
-          }
+        AlignedVector<std::array<Tensor<2, dim>, dim + 1>> jacobians_on_stencil(
+          cell_array.size());
+
+        // Create as many chunks of cells as we have threads and spawn the
+        // work
+        unsigned int work_per_chunk =
+          std::max(std::size_t(1),
+                   (cell_array.size() + MultithreadInfo::n_threads() - 1) /
+                     MultithreadInfo::n_threads());
+
+        // we manually use tasks here rather than parallel::apply_to_subranges
+        // because we want exactly as many loops as we have threads - the
+        // initialization of the loops with FEValues is expensive
```

Comment:
This should answer your question: The reason is that `apply_to_subranges` does dynamic loop scheduling with increasingly smaller subsets. For this first loop, it is simply not worth it because the initialization of `FEValues` is pretty expensive. The other loops are cheaper to initialize so we can tolerate it. Of course, one might still wonder whether the dynamic scheduling is helpful for performance or not (I am sceptical), but I think that is the most easily understandable form.

Reply:
OK! Makes sense! Thanks for the explanation!